api/osf_api.py

```diff
@@ -441,3 +441,155 @@ def create_draft_registration(session, node_id=None, schema_id=None):
+    # NOTE: In the preprint payload record below we are adding a license_record
+    # with year set to 2022 even though in most cases we may be using a license
+    # that does not require the year value. There is a weird bug that occurs on
+    # the Edit Preprint page if the preprint does not have a value for year. If
+    # the year is null then the Edit Preprint page thinks that the preprint has
+    # unsaved changes even if you have made no changes on the form page. There
+    # is a ticket for this issue: ENG-3782.
+    url = '/v2/preprints/'
+    raw_payload = {
+        'data': {
+            'type': 'preprints',
+            'attributes': {
+                'title': title,
+                'description': 'Preprint created via the OSF api',
+                'subjects': [[subject_id]],
+                'license_record': {'year': '2022'},
+                'tags': ['qatest', 'selenium'],
+                'has_coi': False,
+                'has_data_links': 'available',
+                'data_links': ['https://osf.io/'],
+                'has_prereg_links': 'no',
+                'why_no_prereg': 'QA Selenium Testing',
+            },
+            'relationships': {
+                'license': {'data': {'type': 'licenses', 'id': license_id}},
+                'provider': {'data': {'type': 'providers', 'id': provider_id}},
+            },
+        }
+    }
+    return_data = session.post(
+        url=url, item_type='preprints', raw_body=json.dumps(raw_payload)
+    )
+    preprint_node = return_data['data']['id']
+    # Step 2: Create a test file in WaterButler and associate it with the preprint
+    wb_upload_url = '{}/v1/resources/{}/providers/{}/'.format(
+        settings.FILE_DOMAIN, preprint_node, 'osfstorage'
+    )
+    metadata = session.put(
+        url=wb_upload_url,
+        query_parameters={'kind': 'file', 'name': 'OSF Test File.txt'},
+        raw_body={},
+    )
+    # The id value from WB has the provider name and '/' at the beginning of it
+    # (ex: 'osfstorage/627a5b3ab4f587000aa2725c').  So we need to parse out the
+    # actual file id to use in the Preprint patch.
+    file_id = metadata['data']['id'].split('/')[1]
+    # Step 3: Attach the file to the Preprint and set it as Published
+    patch_url = '/v2/preprints/{}/'.format(preprint_node)
+    patch_payload = {
+        'data': {
+            'id': preprint_node,
+            'type': 'preprints',
+            'attributes': {
+                'is_published': True,
+            },
+            'relationships': {
+                'primary_file': {'data': {'type': 'files', 'id': file_id}}
+            },
+        }
+    }
+    return_data = session.patch(
+        url=patch_url,
+        item_type='preprints',
+        item_id=preprint_node,
+        raw_body=json.dumps(patch_payload),
+    )
+    # Return the Preprint Node Id
+    return return_data['data']['id']
+
+
+def get_license_id(session=None, license_name='CC0 1.0 Universal'):
```

Comment:
Oops! My bad! I need to change this to get the license id for a specific provider type and provider id like I do below in get_subject_for_provider.  Each provider has its own list of licenses that it allows. And it should work for both Preprints and Registrations.

Reply:
Refactored function.